pkg/ccl/changefeedccl/cdceval/expr_eval.go

```diff
@@ -0,0 +1,634 @@
+	}()
+	if err != nil {
+		return "", err
+	}
+
+	if as == "" {
+		as = fmt.Sprintf("column_%d", 1+len(e.selectors))
+	}
+	return e.makeUniqueName(as), nil
+}
+
+// makeUniqueName returns a unique name for the specified name.
+// We do this because seeing same named fields in JSON might be confusing.
+func (e *exprEval) makeUniqueName(as string) string {
+	useCount := e.nameUseCount[as]
+	e.nameUseCount[as]++
+	if useCount > 0 {
+		// Unique-ify evalProjection name.
+		as = fmt.Sprintf("%s_%d", as, useCount)
+	}
+	return as
+}
+
+// addSelector adds specified select expression to evalProjection set.
+func (e *exprEval) addSelector(
+	ctx context.Context, selector tree.SelectExpr, numSelectors int,
+) error {
+	as, err := e.computeRenderColumnName(selector)
+	if err != nil {
+		return err
+	}
+
+	typedExpr, err := e.typeCheck(ctx, selector.Expr, types.Any)
+	if err != nil {
+		return err
+	}
+
+	// Expand "*".  We walked expression during type check above, so we only expect to
+	// see UnqualifiedStar.
+	if _, isStar := typedExpr.(tree.UnqualifiedStar); isStar {
+		if numSelectors == 1 {
+			// Single star gets special treatment.
+			e.starProjection = true
+		} else {
+			for ord, col := range e.ResultColumns() {
+				e.addProjection(e.iVarHelper.IndexedVar(ord), e.makeUniqueName(col.Name))
+			}
+		}
+	} else {
+		e.addProjection(typedExpr, as)
+	}
+
+	return nil
+}
+
+// addFilter adds where clause filter.
+func (e *exprEval) addFilter(ctx context.Context, where tree.Expr) error {
+	if where == nil {
+		return nil
+	}
+	typedExpr, err := e.typeCheck(ctx, where, types.Bool)
+	if err != nil {
+		return err
+	}
+
+	if typedExpr == tree.DBoolTrue {
+		if log.V(1) {
+			log.Infof(ctx, "ignoring tautological filter %q", where)
+		}
+		return nil
+	}
+
+	if typedExpr == tree.DBoolFalse {
+		return errors.Newf("filter %q is a contradiction", where)
+	}
+
+	e.filter = typedExpr
+	return nil
+}
+
+// addProjection adds expression to be returned by evalProjection.
+func (e *exprEval) addProjection(expr tree.TypedExpr, as string) {
+	e.selectors = append(e.selectors, expr)
+	e.projection.AddValueColumn(as, expr.ResolvedType())
+}
+
+// typeCheck converts expression to the expression of specified target type.
+func (e *exprEval) typeCheck(
+	ctx context.Context, expr tree.Expr, targetType *types.T,
+) (tree.TypedExpr, error) {
+	// If we have variable free immutable expressions, then we can just evaluate it right away.
+	typedExpr, err := schemaexpr.SanitizeVarFreeExpr(
+		ctx, expr, targetType, "cdc", &e.semaCtx, volatility.Immutable)
+	if err == nil {
+		d, err := eval.Expr(e.evalCtx, typedExpr)
+		if err != nil {
+			return nil, err
+		}
+		return d, nil
+	}
+
+	// We must work harder.  Bind variables and resolve names.
+	expr, _ = tree.WalkExpr(&e.resolver, expr)
+	if e.resolver.err != nil {
+		return nil, e.resolver.err
+	}
+
+	if star, isStar := expr.(tree.UnqualifiedStar); isStar {
+		// Can't type check star -- we'll handle it later during eval.
```

Comment:
Could you use `types.AnyTuple` here?

Reply:
Tbh, I'm not sure.  I see that this stuff is used when generating row(); i'm not sure how this would help here though.
* expansion still happens, and I happen to know exactly what types those expressions are...  Do you have any code gist where you think AnyTuple might come in handy?

Reply:
Thinking about it...probably no point, especially with no aggregate functions.